File: apps/android/app/src/main/java/dev/tohure/android_rust_test/format/MoneyFormatter.kt
package dev.tohure.android_rust_test.format

import java.math.BigDecimal

/**
 * Agrega `S/` y separadores de miles. **Solo al pintar.**
 *
 * Recibe el `String` que devolvió el core y lo convierte a `BigDecimal` —nunca a `Double`—
 * únicamente para agrupar los miles. No redondea: el core ya entregó el valor con la
 * escala correcta (2 decimales para PEN).
 *
 * No se usa en los mensajes de error: ahí los montos van crudos, porque los formateadores
 * de Android, iOS y el navegador no coinciden entre sí.
 */
object MoneyFormatter {
    fun format(amount: String): String {
        val value = runCatching { BigDecimal(amount) }.getOrNull() ?: return amount
        // El signo se separa ANTES de agrupar: si entra al `chunked(3)` se comporta como un
        // dígito más y, cuando la parte entera tiene un múltiplo de 3 dígitos, queda aislado
        // en su propio grupo — "-123456.78" salía como "S/ -,123,456.78".
        val sign = if (value.signum() < 0) "-" else ""
        val parts = value.abs().toPlainString().split(".")
        val grouped = parts[0]
            .reversed()
            .chunked(3)
            .joinToString(",")
            .reversed()
        val decimals = parts.getOrNull(1)
        return if (decimals != null) "S/ $sign$grouped.$decimals" else "S/ $sign$grouped"
    }
}

File: apps/ios/ios-rust-test/Format/MoneyFormatter.swift
import Foundation

/// Agrega `S/` y separadores de miles. **Solo al pintar.**
///
/// Recibe el `String` que devolvió el core y **no redondea**: el core ya entregó el valor
/// con la escala correcta (2 decimales para PEN).
///
/// **No usa `NumberFormatter` a propósito.** Su salida depende del ICU de la plataforma
/// —puede meter un espacio duro entre el símbolo y el número— y no tiene por qué coincidir
/// carácter por carácter con la de Kotlin. La demo pone las cuatro pantallas lado a lado,
/// así que aquí se usa el mismo algoritmo manual que `MoneyFormatter.kt` en Android.
///
/// No se usa en los mensajes de error: ahí los montos van crudos.
enum MoneyFormatter {
    static func format(_ amount: String) -> String {
        // `Decimal(string:locale:)` con POSIX para que el punto decimal se interprete bien
        // sin importar el locale del aparato. NUNCA `Double`. Solo se usa para validar que
        // la entrada es numérica: los decimales que se pintan salen del string original.
        guard Decimal(string: amount, locale: Locale(identifier: "en_US_POSIX")) != nil else {
            return amount
        }

        let sign = amount.hasPrefix("-") ? "-" : ""
        let unsigned = sign.isEmpty ? amount : String(amount.dropFirst())
        let parts = unsigned.split(separator: ".", maxSplits: 1, omittingEmptySubsequences: false)
        guard let integerPart = parts.first else { return amount }

        let grouped = String(
            String(integerPart)
                .reversed()
                .enumerated()
                .map {
                    let isGroupBoundary = $0.offset > 0 && $0.offset.isMultiple(of: 3)
                    return isGroupBoundary ? ",\($0.element)" : "\($0.element)"
                }
                .joined()
                .reversed()
        )

        if parts.count == 2 {
            return "S/ \(sign)\(grouped).\(parts[1])"
        }
        return "S/ \(sign)\(grouped)"
    }
}
